src/nnet3/nnet-optimize.cc

```diff
@@ -845,41 +845,121 @@ void Optimize(const NnetOptimizeOptions &config,
 
 }
 
+// ComputationRequests are distinguished by the names and indexes
+// of inputs and outputs
+size_t ComputationRequestHasher::operator() (const ComputationRequest *cr) const {
+  size_t ans = 0;
+  std::vector<IoSpecification>::const_iterator itr = cr->inputs.begin(),
+                                               end = cr->inputs.end();
+  for (; itr != end; ++itr) {
+    ans += IoSpecificationToInt(*itr);
+  }
+  itr = cr->outputs.begin();
+  end = cr->outputs.end();
+  for (; itr != end; ++itr) {
+    ans += IoSpecificationToInt(*itr);
+  }
+  return ans;
+}
+
+size_t ComputationRequestHasher::IoSpecificationToInt(const IoSpecification& spec) const {
+  size_t ans;
+  StringHasher string_hasher;
+  ans = string_hasher(spec.name);
+  std::vector<Index>::const_iterator itr = spec.indexes.begin(),
+                                     end = spec.indexes.end();
+  for (; itr != end; ++itr) {
+    ans += (*itr).n * 1619;
+    ans += (*itr).t * 15649;
+    ans += (*itr).x * 89809;
+  }
+  return ans;
+}
+
+void CachingOptimizingCompiler::UpdateCache(const ComputationRequest *request,
+                                            NnetComputation *computation) {
+  if (computation_cache_.size() == capacity_) {
+    // full, locate the least-recently-accessed request
+    const typename CacheType::iterator it
+       = computation_cache_.find(access_queue_.front());
+    KALDI_ASSERT(it != computation_cache_.end());
+    // purge the least-recently-accessed request
+    delete it->first;
+    delete it->second.first;
+    computation_cache_.erase(it);
+    access_queue_.pop_front();
+  }
+  typename AqType::iterator ait
+    = access_queue_.insert(access_queue_.end(), request);
+  computation_cache_.insert(std::make_pair(request,
+                            std::make_pair(computation, ait)));
+}
+
+void CachingOptimizingCompiler::UpdateAccessQueue(typename CacheType::iterator &cit) {
+  // exist, update access record by moving the accessed
+  // request to the end of the access queue
+  KALDI_ASSERT(cit != computation_cache_.end());
+  access_queue_.splice(access_queue_.end(), access_queue_,
+                       cit->second.second);
+}
+
+CachingOptimizingCompiler::~CachingOptimizingCompiler() {
+  typename CacheType::const_iterator itr = computation_cache_.begin(),
+                                     end = computation_cache_.end();
+  for (; itr !=end; ++itr) {
+    delete itr->first;
+    delete itr->second.first;
+  }
+  access_queue_.clear();
```

Comment:
I don't think these two clear() commands are really necessary because it's the destructor.  (I know there are instances where existing code does this).